src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs

```diff
@@ -112,43 +120,55 @@ public void Stop()
-            bool overrideExisting)
+            bool overrideExisting,
+            bool isParallelProcessingSupported = false)
         {
             if (overrideExisting)
             {
                 // Remove the existing handler so a new one can be set
                 this.requestHandlers.Remove(requestType.MethodName);
             }
 
+            this.requestHandlerParallelismMap.Add(requestType.MethodName, isParallelProcessingSupported);
             this.requestHandlers.Add(
                 requestType.MethodName,
                 async (requestMessage, messageWriter) =>
                 {
-                    var requestContext =
-                        new RequestContext<TResult>(
-                            requestMessage,
-                            messageWriter);
-                    try
-                    {                        
-                        TParams typedParams = default(TParams);
-                        if (requestMessage.Contents != null)
-                        {
-                            try
+                    Func<Task> func = async () => {
+                        var requestContext =
+                            new RequestContext<TResult>(
+                                requestMessage,
+                                messageWriter);
+                        try
+                        {                        
+                            TParams typedParams = default(TParams);
+                            if (requestMessage.Contents != null)
                             {
-                                typedParams = requestMessage.Contents.ToObject<TParams>();
-                            }
-                            catch (Exception ex)
-                            {
-                                throw new Exception($"Error parsing message contents {requestMessage.Contents}", ex);
+                                try
+                                {
+                                    typedParams = requestMessage.Contents.ToObject<TParams>();
+                                }
+                                catch (Exception ex)
+                                {
+                                    throw new Exception($"Error parsing message contents {requestMessage.Contents}", ex);
+                                }
                             }
+
+                            await requestHandler(typedParams, requestContext);
+                        }
+                        catch (Exception ex)
+                        {
+                            Logger.Error($"{requestType.MethodName} : {ex}");
+                            await requestContext.SendError(ex.Message);
                         }
+                    };
 
-                        await requestHandler(typedParams, requestContext);
+                    if (!this.ParallelMessageProcessing || !isParallelProcessingSupported) {
+                        await func();
```

Comment:
This part is to make sure if a user does not explicitly turn on the `ParallelMessageProcessing` flag, the message handling behavior will remain unchanged.